modules: mbedtls: add helper Kconfig PSA_CRYPTO

valeriosetti opened the pull request:
I found that in the Zephyr codebase there are several occurrences of the following pattern:
```
    select MBEDTLS if !BUILD_WITH_TFM
    select MBEDTLS_PSA_CRYPTO_C if !BUILD_WITH_TFM
```
This is used to enable Mbed TLS' PSA Crypto Core when TF-M is not available in the build. This is not really nice because it brings the PSA Crypto enablement logic down to the subsystem/module/sample/test using it and it also causes code duplication.

The goal of this PR is to introduce `PSA_CRYPTO_PROVIDER` as a solution to this problem. This is meant to simplify the end user life: they will just need to enable `PSA_CRYPTO_PROVIDER` and then Zephyr will take care of enabling what's needed based on the current platform.

valeriosetti committed:
modules: mbedtls: add new helper Kconfig symbol PSA_CRYPTO

The goal of new Kconfig PSA_CRYPTO_PROVIDER is to automatically enable
any of the PSA Crypto API provider available for the platform without
having the user to manually pick the proper one. This provider can be
either TF-M, if that's enabled in the build, or Mbed TLS otherwise.

PSA_CRYPTO_PROVIDER simplifies also modules/subsystem Kconfigs removing
blocks as:
	select MBEDTLS if !BUILD_WITH_TFM
	select MBEDTLS_PSA_CRYPTO_C if !BUILD_WITH_TFM

Kconfig PSA_CRYPTO_PROVIDER_CUSTOM is also added to allow the end user
to add a custom implementation of PSA Crypto API instead of TF-M or
Mbed TLS ones.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
modules: openthread: fix dependency for OPENTHREAD_CRYPTO_PSA

The dependency should be PSA_CRYPTO_CLIENT and not
MBEDTLS_PSA_CRYPTO_CLIENT because the former is more generic. TF-M can
indeed provide PSA Crypto API, not only Mbed TLS.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

asmellby commented on drivers/bluetooth/hci/Kconfig.esp32:
```diff
@@ -487,12 +487,11 @@ config ESP32_BT_LE_CRYPTO_STACK_MBEDTLS
 	default y
-	select MBEDTLS
+	select PSA_CRYPTO_PROVIDER
 	select MBEDTLS_ECP_C
```
Since this driver seems to depend on legacy Mbed TLS crypto APIs, shouldn't it continue to select `MBEDTLS` directly? While it will in practice always be selected through `PSA_CRYPTO_PROVIDER` since this platform doesn't have TF-M (it's not ARM), one could envision swapping to a different PSA implementation but still use legacy Mbed TLS built-in crypto.

valeriosetti replied:
Oh right, I just checked that driver and it seems to me that's completely relying on legacy crypto so I really don't get why they would need `MBEDTLS_PSA_CRYPTO_C` to be enabled in the build. I will revert this change and just remove `MBEDTLS_PSA_CRYPTO_C` from the original configuration. Thanks for the tip

asmellby commented on drivers/bluetooth/hci/Kconfig:
```diff
@@ -158,8 +158,7 @@ config BT_SILABS_EFR32
-	select MBEDTLS
-	select MBEDTLS_PSA_CRYPTO_C
+	select PSA_CRYPTO_PROVIDER
 	select MBEDTLS_ENTROPY_C
```
Similar argument, but opposite conclusion: `MBEDTLS_ENTROPY_C` is actually not required for the Silicon Labs HCI driver -- it just needs `psa_generate_random()`, so this should be removed rather than re-adding an `MBEDTLS` dependency.

asmellby reviewed (commented):
Nice! Looks mostly good to me.

alwa-nordic commented:
The PSA user already signals it wants PSA with the various `select PSA_WANT_` statements. I suggest instead of the user selecting `PSA_CRYPTO_PROVIDER` manually, it should be automatically *defaulted* (not selected) by the all the `PSA_WANT_` configs.

We can also introduce a set of `PSA_NEED_` configs which will *select* `PSA_CRYPTO_PROVIDER`.

My scheme simplifies the usage (Just select features you want or need), you can *want* in order to suggest the presence of specific PSA features, and you can *need* in order to require features.

valeriosetti committed:
drivers: bluetooth: esp32: remove selection of MBEDTLS_PSA_CRYPTO_C

The driver code only relies on legacy Mbed TLS crypto, not on PSA API, so
enabling MBEDTLS_PSA_CRYPTO_C is not needed here.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
drivers: bluetooth: hci: do not select MBEDTLS_ENTROPY_C in BT_SILABS_EFR32

The driver only uses psa_generate_random() so ENTROPY_C is not required.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti commented on modules/mbedtls/Kconfig.psa.logic:
```diff
@@ -1,8 +1,37 @@
+	  Enable a PSA Crypto API provider in the build. If TF-M is enabled then
+	  it will be used for this scope, otherwise Mbed TLS will be used.
+
+choice PSA_CRYPTO_PROVIDER_IMPL
```
Yeah the problem here arose from the fact that the base Kconfig already had `_PROVIDER` suffix so I had to come up with a double suffix (`_PROVIDER_IMPL`) which I agree it's not nice.
@frkv suggested to rename `PSA_CRYPTO_PROVIDER` as just `PSA_CRYPTO` so I think this would resolve this problem as well because then I can use a single prefix.

valeriosetti commented on modules/mbedtls/Kconfig.psa.logic:
```diff
@@ -1,8 +1,37 @@
+choice PSA_CRYPTO_PROVIDER_IMPL
+	prompt "PSA Crypto API provider options"
+	depends on PSA_CRYPTO_PROVIDER
+	default PSA_CRYPTO_PROVIDER_TFM if BUILD_WITH_TFM
+	default PSA_CRYPTO_PROVIDER_MBEDTLS
```
For the TF-M case I agree, but for the non-TF-M case I think it still makes sense to have a preference between `PSA_CRYPTO_PROVIDER_MBEDTLS` and `PSA_CRYPTO_PROVIDER_CUSTOM`, wdyt?

valeriosetti replied:
Done :wink:

asmellby reviewed (commented):
The commit message of the last commit should be updated to reflect the removal of `_PROVIDER` from the public-facing option name. Other than that, looks great!

Vge0rge reviewed (dismissed):
Very nice improvement :)

tomi-font commented on modules/mbedtls/Kconfig.psa.logic:
```diff
@@ -1,8 +1,35 @@
+	  Enable a PSA Crypto API provider in the build. If TF-M is enabled then
+	  it will be used for this scope, otherwise Mbed TLS will be used.
+
+choice PSA_CRYPTO_PROVIDER
+	prompt "PSA Crypto API provider"
+	depends on PSA_CRYPTO
```
How about this selects `PSA_CRYPTO_CLIENT`?

alxelax commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -1501,8 +1501,7 @@ choice BT_MESH_CRYPTO_LIB
 	bool "mbed TLS PSA"
-	select MBEDTLS
-	select MBEDTLS_PSA_CRYPTO_C
+	select PSA_CRYPTO
```
I'm not sure if this is correct change, since crypto library is not selected via `PSA_CRYPTO_PROVIDER` choice in Mesh.

With this change, it is possible to choose both `BT_MESH_USES_MBEDTLS_PSA` and `PSA_CRYPTO_PROVIDER_TFM` at the same time.
I even cannot imagine to what it might cause.

Mesh KConfig options should be reworked to use `PSA_CRYPTO_PROVIDER` correctly according to designed pattern.
Or shouldn't be changed at all.

This semicorrect variant is not the option.

valeriosetti committed:
bluetooth: mesh: use new Kconfig CONFIG_PSA_CRYPTO

Remove previous Kconfig choices CONFIG_BT_MESH_USES_MBEDTLS_PSA and
CONFIG_BT_MESH_USES_TFM_PSA and use CONFIG_PSA_CRYPTO instead.
This commit also updates test code accordingly.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
doc: releases: add note about CONFIG_PSA_CRYPTO addition in Mbed TLS

Update release-notes about the following additions:
- CONFIG_PSA_CRYPTO
- CONFIG_PSA_CRYPTO_PROVIDER_TFM
- CONFIG_PSA_CRYPTO_PROVIDER_MBEDTLS
- CONFIG_PSA_CRYPTO_PROVIDER_CUSTOM

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti commented:
@ceolin @d3zd3z can you please take a look?

ceolin reviewed (approved):
it looks great !

frkv reviewed (approved):
LGTM. May perhaps select `MBEDTLS_PSA_CRYPTO_CLIENT` from setting a crypto provider, but that is OK to handle in a follow-up work

tomi-font commented:
> Isn't PSA_CRYPTO_CLIENT something that should be selected by the provider (i.e. the real implementation)? A similar thing happens for driver/entropy: the user selects CONFIG_ENTROPY_GENERATOR and then it's the driver that selects CONFIG_ENTROPY_HAS_DRIVER.

The analogy you are making here to me doesn't apply. This mechanism is needed for the entropy drivers because we actually don't know if there is an entropy driver available, so the driver needs to manifest itself by selecting `CONFIG_ENTROPY_HAS_DRIVER`.
In the case of `CONFIG_PSA_CRYPTO` we know (expect) that in all cases a PSA Crypto implementation will be enabled, thus we don't have to make every provider enable `PSA_CYPTO_CLIENT` individually. It's just unnecessary.
